Add OpenAI Provider

utkarsharma2 opened the pull request:
This PR is part of our larger effort to add first-class integrations to support LLMOps that was [presented at Airflow Summit](https://www.youtube.com/watch?v=mgA6m3ggKhs&t=4s). 

This PR adds explicitly the OpenAI Provider. OpenAI is a leading American artificial intelligence organization, which offers one of the most used LLM - ChatGPT and offers embedding models. 
 
The primary objective of this Provider is to present users with an alternative embedding model. This allows them to generate vectors for their proprietary data, a pivotal step towards establishing integrations with LLM models like ChatGPT.

Example DAG:
The `OpenAIEmbeddingOperator` can accept either a string or a callable returning a list of strings.
```
OpenAIEmbeddingOperator(
        task_id="embedding_using_xcom_data",
        conn_id="openai_default",
        input_text=xcom_text["input_text"],
        model="text-embedding-ada-002",
    )
```

Email Discussion related to the effort can be found here - https://lists.apache.org/thread/0d669fmy4hn29h5c0wj0ottdskd77ktp

utkarsharma2 committed:
Add OpenAI Provider

phanikumv commented on airflow/providers/openai/hooks/openai.py:
```diff
@@ -0,0 +1,89 @@
+
+    def create_completions(self, prompt: str, model: str = "davinci-002", **kwargs: Any) -> str:
+        """
+        Create a completions for the given text using the given model.
```
```suggestion
        Create completions for the given text using the given model.
```

phanikumv commented on airflow/providers/openai/operators/openai.py:
```diff
@@ -0,0 +1,86 @@
+    Operator that accepts input text to generate OpenAI embeddings using the specified model.
+
+    :param conn_id: The OpenAI connection.
+    :param input_text: The text to generate OpenAI embeddings on. Only one of input_text or input_callable
```
```suggestion
    :param input_text: The text to generate OpenAI embeddings on. Either input_text or input_callable
```

phanikumv commented on airflow/providers/openai/operators/openai.py:
```diff
@@ -0,0 +1,86 @@
+    :param input_text: The text to generate OpenAI embeddings on. Only one of input_text or input_callable
+        should be provided.
+    :param input_callable: The callable that provides the input text to generate OpenAI embeddings.
+        Only one of input_text or input_callable should be provided.
```
```suggestion
        Either input_text or input_callable should be provided.
```

Taragolis commented on airflow/providers/openai/hooks/openai.py:
```diff
@@ -0,0 +1,89 @@
+    def __init__(self, conn_id: str = default_conn_name, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+        self.conn_id = conn_id
+        openai.api_key = self._get_api_key()
```
I don't look other code yet but want to mention that current `openai` library implementation doesn't work well in multiple threads: https://github.com/openai/openai-python/discussions/631

So maybe we should wait a bit and wait until [1.0.0 release](https://github.com/openai/openai-python/discussions/631) and fix this issue, as well as introduce breaking changes

There are also some other issues currently exists, which I hope fixed into the new major version of library.

utkarsharma2 replied:
@Taragolis I see there are a bunch of issues, what if release a minor/patch version of this provider as well? Just to indicate this is not production-ready?

cc: @kaxil @mpgreg

mpgreg replied:
Yes let's get it out with a note on status.

Taragolis replied:
And maybe also better to pin openai library <1.0

utkarsharma2 replied:
Made the required changes, please check.

phanikumv commented on airflow/providers/openai/hooks/openai.py:
```diff
@@ -0,0 +1,89 @@
+
+    .. seealso:: https://platform.openai.com/docs/introduction/overview
+
+    :param conn_id: :ref:`Cohere connection id <howto/connection:openai>`
```
```suggestion
    :param conn_id: :ref:`OpenAI connection id <howto/connection:openai>`
```

pankajkoti committed:
Apply suggestions from code review

Co-authored-by: Phani Kumar <94376113+phanikumv@users.noreply.github.com>

mpgreg commented on airflow/providers/openai/hooks/openai.py:
```diff
@@ -0,0 +1,89 @@
+        return str(conn.password)
+
+    def create_embeddings(
+        self, text: str, model: str = "text-embedding-ada-002", **kwargs: Any
```
The client should accept a single string, array of strings or array of tokens per https://platform.openai.com/docs/api-reference/embeddings/create#embeddings/create-input

utkarsharma2 replied:
Sure, I have updated with the changes requested.

mpgreg commented on airflow/providers/openai/hooks/openai.py:
```diff
@@ -0,0 +1,89 @@
+        embeddings: list[float] = response["data"][0]["embedding"]
+        return embeddings
+
+    def create_completions(self, prompt: str, model: str = "davinci-002", **kwargs: Any) -> str:
```
Completions should accept string or array per https://platform.openai.com/docs/api-reference/completions/create#completions/create-prompt

utkarsharma2 replied:
Since we don't plan on exposing the operator for it as of now. I'll remove this method.

mpgreg commented on airflow/providers/openai/operators/openai.py:
```diff
@@ -0,0 +1,86 @@
+    def __init__(
+        self,
+        conn_id: str,
+        input_text: str | None = None,
```
input_text should be str | list | None.

utkarsharma2 replied:
Sure, updated the changes requested.

mpgreg commented on airflow/providers/openai/provider.yaml:
```diff
@@ -0,0 +1,52 @@
+    tags: [software]
+
+dependencies:
+  - apache-airflow>=2.5.0
```
Is there something inherent in 2.5 minimum?  I didn't see any dependency collisions.

utkarsharma2 replied:
It's not a technical dependency. Based on [this](https://github.com/apache/airflow/blob/main/PROVIDERS.rst#minimum-supported-version-of-airflow-for-community-managed-providers) document I assume, all new provider releases from 2nd of December 2022 to 2nd December will have a min version as 2.5.0.

mpgreg commented on docs/apache-airflow-providers-openai/index.rst:
```diff
@@ -0,0 +1,95 @@
+Requirements
+------------
+
+The minimum Apache Airflow version supported by this provider package is ``2.4.0``.
```
provider.yml lists minimum 2.5.0

utkarsharma2 replied:
Updated to 2.5.0 because of same reason as https://github.com/apache/airflow/pull/35023#discussion_r1368622509

mpgreg commented on docs/apache-airflow-providers-openai/index.rst:
```diff
@@ -0,0 +1,95 @@
+==================  ==================
+PIP package         Version required
+==================  ==================
+``apache-airflow``   ``>=2.5.0``
```
2.5 or 2.4?

utkarsharma2 replied:
Commented here - https://github.com/apache/airflow/pull/35023#discussion_r1368622509

mpgreg commented on generated/provider_dependencies.json:
```diff
@@ -647,6 +647,14 @@
   },
+  "openai": {
+    "deps": [
+      "apache-airflow>=2.5.0",
```
2.5 or 2.4?

utkarsharma2 replied:
Commented here - https://github.com/apache/airflow/pull/35023#discussion_r1368622509

mpgreg commented on airflow/providers/openai/operators/openai.py:
```diff
@@ -0,0 +1,86 @@
+        self.conn_id = conn_id
+        self.input_text = input_text
+        self.input_callable = input_callable
+        self.input_callable_args = input_callable_args or ()
```
Similar to discussion for cohere and pgvector its a bit awkward to provide these args.  The point is to allow users to ingest without passing data through xcoms.  The awkward passing of args could be alleviated by a decorator as discussed at https://github.com/apache/airflow/pull/34891#discussion_r1365231569

utkarsharma2 replied:
I have changed the interface, now we just have one templated field `input_text`. I think it addresses all of the concerns.

utkarsharma2 committed:
Remove create_completions method from hook

utkarsharma2 committed:
Change type of input_text param

Since the upstream API accepts str ot list of tokens, we accept the similar inputs from user.

utkarsharma2 committed:
Updated min-airflow version to 2.5.0

utkarsharma2 committed:
Updated the interface and fix docs and static files

utkarsharma2 committed:
Fix tests

utkarsharma2 committed:
Fix tests

hussein-awala commented on airflow/providers/openai/operators/openai.py:
```diff
@@ -0,0 +1,73 @@
+        **kwargs: Any,
+    ):
+        self.embedding_params = kwargs.pop("embedding_params", {})
+        self.hook_params = kwargs.pop("hook_params", {})
```
It would be better to add them to params (and docstring), then use:
```python
self.embedding_params = embedding_params or {}
self.hook_params = hook_params or {}
```
otherwise the user should read the source code to know about these params.

utkarsharma2 replied:
@hussein-awala I agree, that we should be more explicit, but here we are just passing the `kwargs` to BaseHook `super().__init__(*args, **kwargs)` and not the OpenAI client. But I'm happy to remove 
`self.hook_params = kwargs.pop("hook_params", {})` and related logic if that is not the ideal way of doing it.

Also,  WRT `self.embedding_params = kwargs.pop("embedding_params", {})` I think if we are explicit with all the params supported with `openai.Embedding.create` method, we may have to do releases every time there is a change with respect to params in OpenAI SDK.

Maybe we can take a middle ground here and in our operator params explicitly mention the possible options for `embedding_params` by referring to the link `https://platform.openai.com/docs/api-reference/embeddings` as OpenAI does in there SDK. So it can be immune to changes in  OpenAI SDK. WDYT?

pankajkoti replied:
maybe the suggestion here is to accept embedding_params and hook_params explicitly as operator params instead of it being hidden under kwargs. So that it gets reflected in the docs. but let's wait for @hussein-awala to confirm.

utkarsharma2 replied:
Sorry, I misread it. I'll update based on @pankajkoti comment.

utkarsharma2 replied:
Removed the `hook_params` as it was not getting used and added `embedding_params` as a param.

utkarsharma2 committed:
Change the version

Because of OpenAI SDK not being production ready

utkarsharma2 commented on CONTRIBUTING.rst:
```diff
@@ -676,10 +676,10 @@ dbt.cloud, deprecated_api, devel, devel_all, devel_ci, devel_hadoop, dingding, d
+microsoft.winrm, mongo, mssql, mysql, neo4j, odbc, openai, openfaas, openlineage, opensearch,
+opsgenie, oracle, otel, pagerduty, pandas, papermill, password, pinot, plexus, postgres, presto,
+rabbitmq, redis, s3, s3fs, salesforce, samba, segment, sendgrid, sentry, sftp, singularity, slack,
+smtp, snowflake, spark, sqlite, ssh, statsd, tableau, tabular, telegram, trino, vertica, virtualenv,
```
```suggestion
microsoft.winrm, mongo, mssql, mysql, neo4j, odbc, openai, openfaas, openlineage, opensearch, oracle, otel, pagerduty, pandas, papermill, password, pinot, plexus, postgres, presto, redis, s3, s3fs, salesforce, samba, segment, sendgrid, sentry, sftp, singularity, slack, snowflake, spark, sqlite, ssh, statsd, tableau, tabular, telegram, trino, vertica, virtualenv,
```

utkarsharma2 commented on INSTALL:
```diff
@@ -103,10 +103,10 @@ dbt.cloud, deprecated_api, devel, devel_all, devel_ci, devel_hadoop, dingding, d
+microsoft.winrm, mongo, mssql, mysql, neo4j, odbc, openai, openfaas, openlineage, opensearch,
+opsgenie, oracle, otel, pagerduty, pandas, papermill, password, pinot, plexus, postgres, presto,
+rabbitmq, redis, s3, s3fs, salesforce, samba, segment, sendgrid, sentry, sftp, singularity, slack,
+smtp, snowflake, spark, sqlite, ssh, statsd, tableau, tabular, telegram, trino, vertica, virtualenv,
```
```suggestion
microsoft.winrm, mongo, mssql, mysql, neo4j, odbc, openai, openfaas, openlineage, opensearch, oracle, otel, pagerduty, pandas, papermill, password, pinot, plexus, postgres, presto, redis, s3, s3fs, salesforce, samba, segment, sendgrid, sentry, sftp, singularity, slack, snowflake, spark, sqlite, ssh, statsd, tableau, tabular, telegram, trino, vertica, virtualenv,
```

utkarsharma2 committed:
Add embedding_kwargs as a param to operator

pankajastro commented on airflow/providers/openai/CHANGELOG.rst:
```diff
@@ -0,0 +1,24 @@
+Changelog
+---------
+
+0.0.1
```
```suggestion
1.0.0
```

pankajastro commented on airflow/providers/openai/operators/openai.py:
```diff
@@ -0,0 +1,75 @@
+        should be provided.
+    :param input_callable: The callable that provides the input text to generate OpenAI embeddings.
+        Either input_text or input_callable should be provided.
+    :param input_callable_args: The list of arguments to be passed to ``input_callable``
```
we should remove the param which are not in init

utkarsharma2 replied:
Oops, missed it. Removed now.

pankajastro commented on airflow/providers/openai/operators/openai.py:
```diff
@@ -0,0 +1,75 @@
+    :param input_callable: The callable that provides the input text to generate OpenAI embeddings.
+        Either input_text or input_callable should be provided.
+    :param input_callable_args: The list of arguments to be passed to ``input_callable``
+    :param input_callable_kwargs: The kwargs to be passed to ``input_callable``
```
same

utkarsharma2 replied:
Oops, missed it. Removed now.

pankajastro commented on airflow/providers/openai/provider.yaml:
```diff
@@ -0,0 +1,54 @@
+suspended: false
+
+versions:
+  - 0.0.1
```
```suggestion
  - 1.0.0
```

utkarsharma2 replied:
@pankajastro We need to have `0.0.1` since the SDK of OpenAI is not production-ready.

eladkal replied:
Airflow provider version is not correlated with SDK version
We have never released 0.0.1 provider.
i don't think there is a problem with doing that (cc @potiuk ?)
but as a general rule don't try to create binding between the two.

kaxil replied:
Starting with 0.0.1 is odd though, “patch” version of an unreleased package :) .

0.1.0 or 1.0.0 is fine for me

If OpenAI releases a breaking change, we can bump up the major version.

utkarsharma2 replied:
Changed the version back to 1.0.0.

pankajastro commented on tests/providers/openai/hooks/test_openai.py:
```diff
@@ -0,0 +1,63 @@
+
+from airflow.providers.openai.hooks.openai import (
+    OpenAIHook,
+)
```
```suggestion
from airflow.providers.openai.hooks.openai import OpenAIHook
```

pankajastro reviewed (commented):
same as https://github.com/apache/airflow/pull/34921#issuecomment-1790420816

utkarsharma2 committed:
Update tests/providers/openai/hooks/test_openai.py

Co-authored-by: Pankaj Singh <98807258+pankajastro@users.noreply.github.com>

utkarsharma2 committed:
Remove unwanted params in docstring

utkarsharma2 committed:
Update Changelog

utkarsharma2 committed:
Add security.rst file

pankajastro commented on docs/apache-airflow-providers-openai/index.rst:
```diff
@@ -0,0 +1,95 @@
+    :maxdepth: 1
+    :caption: Basics
+
+    Home <self>
```
```suggestion
    Home <self>
    Security <security>
```

utkarsharma2 committed:
Update docs/apache-airflow-providers-openai/index.rst

Co-authored-by: Pankaj Singh <98807258+pankajastro@users.noreply.github.com>

utkarsharma2 committed:
Add host field for connections

pankajastro commented on docs/apache-airflow-providers-openai/index.rst:
```diff
@@ -0,0 +1,96 @@
+    :maxdepth: 1
+    :caption: Basics
+
+    Home <self>
```
```suggestion
    Home <self>
    Changelog <changelog>
```

utkarsharma2 committed:
Update docs/apache-airflow-providers-openai/index.rst

Co-authored-by: Pankaj Singh <98807258+pankajastro@users.noreply.github.com>

utkarsharma2 committed:
Add changelog.rst file to docs

utkarsharma2 committed:
Change version to 1.0.0

pankajastro reviewed (approved):
LGTM

utkarsharma2 committed:
Resolve conflicts

utkarsharma2 committed:
Fix tests

utkarsharma2 committed:
Fixed tests

utkarsharma2 committed:
Fix test

utkarsharma2 committed:
Resolve Conflict
